src/theory/strings/theory_strings_rewriter.cpp

```diff
@@ -3008,9 +3009,27 @@ Node TheoryStringsRewriter::rewriteStrConvert(Node node)
       newChar = CVC4::String::convertCodeToUnsignedInt(newChar);
       nvec[i] = newChar;
     }
-    Node retNode = NodeManager::currentNM()->mkConst(String(nvec));
+    Node retNode = nm->mkConst(String(nvec));
     return returnRewrite(node, retNode, "str-conv-const");
   }
+  else if (node[0].getKind() == STRING_CONCAT)
+  {
+    std::vector<Node> cc;
+    for (const Node& nc : node[0])
+    {
+      cc.push_back(nm->mkNode(nk, nc));
+    }
+    Node retNode = nm->mkNode(STRING_CONCAT, cc);
+    return returnRewrite(node, retNode, "str-conv-minscope-concat");
+  }
```

Comment:
Another normalization that we should do is to decide between `str.tolower(str.substr(x, n, m))` and `str.substr(str.tolower(x), n, m)`. Pushing `str.tolower` out is maybe better because you get `m` as a bound for the quantifier of `str.tolower`?

Reply:
This is tricky, pushing to inner also has advantages: say I have multiple `tolower(substr(x,i1,n1)) ... tolower(substr(x,im,nm))`, then pushing inwards makes it so that I only have a single application of `tolower`.